segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java

```diff
@@ -709,16 +728,30 @@ private void rollover() {
             Ledgers.close(oldLedger);
             log.debug("{}: Rollover: swapped ledger and metadata pointers (Old = {}, New = {}) and closed old ledger.",
                     this.traceObjectId, oldLedger.getId(), newLedger.getId());
-        } catch (Throwable ex) {
-            if (!ExceptionHelpers.mustRethrow(ex)) {
-                log.error("{}: Rollover failure; log may be unusable.", ex);
-            }
+        } finally {
+            // It's possible that we have writes in the queue that didn't get picked up because they exceeded the predicted
+            // ledger length. Invoke the Write Processor to execute them.
+            this.writeProcessor.runAsync();
+            LoggerHelpers.traceLeave(log, this.traceObjectId, "rollover", traceId, true);
         }
+    }
 
-        // It's possible that we have writes in the queue that didn't get picked up because they exceeded the predicted
-        // ledger length. Invoke the Write Processor to execute them.
-        this.writeProcessor.runAsync();
-        LoggerHelpers.traceLeave(log, this.traceObjectId, "rollover", traceId, true);
+    /**
+     * Handles a failure from the Rollover Processor.
+     *
+     * @param exception    The causing exception.
+     * @param failureCount The number of consecutive failures.
+     * @return True if the RolloverProcessor should be reinvoked, false otherwise.
+     */
+    private boolean handleRolloverFailure(Throwable exception, int failureCount) {
+        log.error("{}: Rollover failure (attempt {}/{}); log may be unusable.", this.traceObjectId, failureCount, this.config.getMaxWriteAttempts(), exception);
+        if (failureCount >= this.config.getMaxWriteAttempts()) {
+            log.warn("{}: Too many rollover failures; closing.", this.traceObjectId);
+            close();
```

Comment:
Please clarify one this for me here. What's the ultimate consequence of closing the `BookKeeperLog`? I can see that `BookKeeperLog.close` closes the write and rollover processors, the write queue, and any remaining open ledger. What happens to the container, though? Does it instantiate `BookKeeperLog` again? Does it kill itself?

Reply:
BookKeeperLog should close when it did its best at its job. When it cannot do that anymore (whether it was fenced out or it has retried sufficient times) it will consider that all hope is lost, give up, and close. The Container will detect this via an `ObjectClosedException` and will shut down itself as well, which will give the Controller/ZKSegmentContainerManager a chance to restart it fresh.